check_dns.pl

```diff
@@ -128,101 +136,104 @@ BEGIN
+        } elsif($rr->type eq "NS"){
+            $result = $rr->nsdname;
+        } elsif($rr->type eq "PTR"){
+            $result = $rr->ptrdname;
+        } elsif($rr->type eq "SOA"){
+            $result = $rr->serial;
+        } elsif($rr->type eq "SRV"){
+            $result = $rr->target;
+        } elsif($rr->type eq "TXT"){
+            $result = $rr->txtdata;
+        } else {
+            quit "UNKNOWN", "unknown/unsupported record type '$rr->type' returned for record '$records[$i]'";
+        }
+        vlog2 "got result: $result\n";
+        if($type eq "A"){
+            isIP($result) or quit "CRITICAL", "invalid result '$result' returned for A record by DNS server, expected IP address for A record$perfdata";
+        } elsif(grep { $type eq $_ } qw/CNAME MX NS PTR SRV/){
+            isFqdn($result) or quit "CRITICAL", "invalid result '$result' returned " . ($verbose ? "for record '$record' type '$type' ": "") . "by DNS server, expected FQDN for this record type$perfdata";
+        } elsif($type eq "SOA"){
+            isInt($result) or quit "CRITICAL", "invalid serial result '$result' returned for SOA record " . ($verbose ? "'$records[$i]' ": "") . "by DNS server, expected an unsigned integer$perfdata";
+        }
+        push(@results, $result);
+        if(@expected_results){
+            unless(grep(lc $_ eq lc $result, @expected_results)){
+                vlog3 "result '$result' wasn't found in expected results, added to rogue list";
+                push(@rogue_results, $result);
+            }
         }
     }
-}
-
-@results or quit "CRITICAL", "no result received for '$record' $type record from servers " . join(",", @servers) . " in $total_time secs";
+    @results or quit "CRITICAL", "no result received for '$records[$i]' $type record from servers " . join(",", @servers) . " in $call_time secs";
+    my @results_uniq =  sort(uniq_array(@results));
 
-my @results_uniq = sort(uniq_array(@results));
 
-foreach my $expected_result2 (@expected_results){
-    unless(grep(lc $_ eq lc $expected_result2, @results_uniq)){
-        vlog3 "$expected_result2 wasn't found in results, adding to missing list";
-        push(@missing_results, $expected_result2);
+    foreach my $exp_result (@expected_results){
+        unless(grep(lc $_ eq lc $exp_result, @results_uniq)){
+            vlog3 "$exp_result wasn't found in results, adding to missing list";
+            push(@missing_results, $exp_result);
+        }
     }
-}
 
-my @regex_mismatches;
-if($expected_regex2){
-    foreach my $result (@results){
-        $result =~ /^$expected_regex2$/ or push(@regex_mismatches, $result);
+    my @regex_mismatches;
+    if($valid_expected_regex){
+        foreach my $result (@results){
+            $result =~ /^$valid_expected_regex$/ or push(@regex_mismatches, $result);
+        }
     }
+    @regex_mismatches = sort(uniq_array(@regex_mismatches)) if(@regex_mismatches);
+    $msg .= "\n$records[$i] $type record ";
```

Comment:
multi-line will break some monitoring servers output / dashboards as traditionally only 1 line is guaranteed to be displayed, you wouldn't want the output on some systems to be "DNS OK:" and nothing else...